Add all ledgers endpoint

robertDurst opened the pull request:
Modeled heavily off of All Assets, this endpoint represents the all ledgers endpoint for the stellar horizon server.

Since this is my first PR please critique mercilessly :)

Closes #57

kbacha commented on client/src/endpoint/ledger.rs:
````diff
@@ -0,0 +1,151 @@
+/// ```
+#[derive(Debug, Default)]
+pub struct AllLedgers {
+    // // TODO: Cursor needs to be parseable from the links that come back
````
I may have left this comment in the AllAssets but I think it can be removed now 😄

robertDurst commented:
Some tests failed, I will fix these up.

correlator commented:
we also use cargo fmt to lint the repo.  Travis will also fail if there is a diff between your commit and the cargo fmt output. https://github.com/rust-lang-nursery/rustfmt <-- just a heads up.  Again thanks for joining and taking the initiative to dive right in.  This is awesome!

kbacha commented:
Thanks for jumping in with two feet! This is a great start :)

robertDurst commented:
Hey so this seems to be the problem I am having: 

`thread 'main' panicked at 'called Result::unwrap() on an Err value: ParseError(ErrorImpl { code: Message("missing field `prev_hash`"), line: 51, column: 7 })'`

And it is related to the doc tests.

robertDurst commented:
So it appears prev_hash is not actually a value on a ledger, so that was causing tests to fail. Everything passes now!

https://horizon-testnet.stellar.org/ledgers?limit=1

kbacha commented on resources/src/ledger.rs:
```diff
@@ -37,11 +36,6 @@ impl Ledger {
     }
 
-    /// The hash of the ledger that chronologically came before this one.
-    pub fn prev_hash(&self) -> &String {
```
:+1:

kbacha commented:
@robertDurst I updated some of the ways I documented and arranged the modules in #69 if you want to take a look 😄 

Hopefully those changes make sense and we can match these PRs. I don't mind refactoring yours after we merge though.

robertDurst commented:
Sounds good! I'll check out this stuff and squash tomorrow!

robertDurst commented:
Ok.... let me squash and this should be good to go.

robertDurst commented:
Hey @kbacha so I am not very good at rebasing and squashing 😅 ... I did it once and I'm still not 100% sure I didn't mess it all up.

Essentially here, I assume I would need to move all my commits into one, after the two PR's that I pulled (and were submitted since this PR has been open).

Otherwise I reformatted the ledger docs and I think it all looks good.

correlator commented:
Try following these steps @robertDurst :

1. Checkout and pull master from GitHub
2. checkout robertDurst:all-ledgers
3. `git rebase master`.  From the looks of it you will have conflicts to resolve.  You can also do `git merge master` here if you are more comfortable resolving conflicts in a merge commit.  Conflict resolution in a rebase is slightly different, so up to you on how many new things you want to try at once.

Now you should be on robertDurst:all-ledgers and be up to date with the changes in master.

This is the most explicit way I know of to squash old commits, there are fancier shortcuts out there you can learn about.
4. type `git status`, git log` and you should see your full commit history
5. `git reset HEAD^ --soft`, `git status`, `git log` and you should see that we just undid the previous commit with all changes still staged.
6. Repeat step 5 until you have all the changes you want stagged for commit.  `git commit` should open a vim editor.  The first line should be the title of your commit followed by a few sentences describing with you did.
7. All your changes are saved with a single commit :beers:

kbacha commented:
You may want to use interactive rebase.

https://robots.thoughtbot.com/git-interactive-rebase-squash-amend-rewriting-history#squash-commits-together

robertDurst commented:
Thanks @correlator and @kbacha! So I have squashed all down to one commit.

However it also squashed the merged code I had in the middle of my PR's...

kbacha commented:
I'm guessing that your merge commit may have pulled in some in-between state. You can probably revert those files by checking them out individually.

```bash
$ git checkout CONTRIBUTING.md
```

And then amending your commit:

```bash
$ git commit --amend
$ git push -f origin HEAD
```

kbacha commented:
btw, I'll merge this as soon as those extraneous changes are removed 👍

robertDurst commented:
Learn something new about git everyday... thanks for your patience 👍

robertDurst commented:
Now it is messed up again. Ill give this another go after work.

kbacha commented:
i may have made the checkout harder to understand, but if you do:
```bash
git checkout master -- <filename/directory>
```

then I think it works. Basically, this tells git not to checkout the branch but to checkout a particular file instead.

robertDurst commented:
Got it down to two commits.

robertDurst commented:
Started celebrating when rebase worked... and then I failed the tests. However, I am unable to reproduce this failure @kbacha. Help would be awesome :)

robertDurst commented:
Rebase worked 👍  but then build broke 👎 . Having issues reproducing the error. Any help here would be awesome!

kbacha commented:
@robertDurst I'll rebuild. I wonder if the horizon test api blocked us? It looks like requests timed out.

kbacha commented:
Thanks @robertDurst !

robertDurst commented:
![https://media.giphy.com/media/uudzUtVcsLAoo/giphy.gif](https://media.giphy.com/media/uudzUtVcsLAoo/giphy.gif)
